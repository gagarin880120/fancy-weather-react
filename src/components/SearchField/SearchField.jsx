import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styles from './SearchField.module.css';

export default function SearchField({ onSearch, currentDateInterval }) {
  const [query, setQuery] = useState('');
  return (
    <div className={styles.wrapper}>
      <input
        type="search"
        name="search"
        className={styles.searchInput}
        onChange={(e) => {
          setQuery(e.target.value);
        }}
        onKeyDown={(e) => {
          if (e.key === 'Enter') {
            onSearch(query, currentDateInterval);
          }
        }}
      />
      <button
        type="button"
        onClick={() => onSearch(query, currentDateInterval)}
      >
        Search
      </button>
    </div>
  );
}

SearchField.propTypes = {
  onSearch: PropTypes.func,
  currentDateInterval: PropTypes.number,
}

SearchField.defaultProps = {
  onSearch: null,
  currentDateInterval: 0,
}
