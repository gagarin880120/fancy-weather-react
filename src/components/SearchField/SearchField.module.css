.searchInput {
  height: 35px;
  border: 2px solid #fff;
  background-color: transparent;
  color: #fff;
  border-radius: 5px;
  font-size: 22px;
  outline: 0;
}

.button {
  width: 35px;
  height: 35px;
  border: 2px solid #fff;
  background-color: transparent;
  color: #fff;
  border-radius: 5px;
  font-size: 22px;
  margin-left: 5px;
  cursor: pointer;
  outline: 0;
}

.button:active {
  transform: translateY(2px);
}

.pulsatingCircle {
  position: absolute;
  display: none;
  border: 3px solid rgb(78, 253, 253);
  border-radius: 100%;
  height: 60px;
  width: 60px;
  position: absolute;
  right: 10px;
  top: -17px;
  animation: pulsate .7s ease-out;
  animation-iteration-count: infinite;
  opacity: 0.0
}

@keyframes pulsate {
    0% {transform: scale(0.1, 0.1); opacity: 0.0;}
    50% {opacity: 1.0;}
    100% {transform: scale(1.2, 1.2); opacity: 0.0;}
}

@media screen and (max-width: 420px) {
  .searchInput {
    width: 190px;
  }
}
