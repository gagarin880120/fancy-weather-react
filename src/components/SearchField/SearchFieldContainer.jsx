import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import SearchField from './SearchField';
import { selectCurrentDateInterval, selectLanguage } from '../../redux/selectors';
import { getAddressBySearch, setQuery } from '../../redux/actions';

export function SearchFieldContainer({ onSearch, currentDateInterval, language }) {
  return (
    <SearchField
      onSearch={onSearch}
      currentDateInterval={currentDateInterval}
      language={language}
    />
  );
}

export const mapStateToProps = (state) => ({
  currentDateInterval: selectCurrentDateInterval(state),
  language: selectLanguage(state),
});

export const mapDispatchToProps = (dispatch) => ({
  onSearch(query, interval, lang) {
    dispatch(setQuery(query));
    dispatch(getAddressBySearch(query, lang, false, interval));
  },
});

SearchFieldContainer.propTypes = {
  onSearch: PropTypes.func,
  currentDateInterval: PropTypes.number,
  language: PropTypes.string,
};

SearchFieldContainer.defaultProps = {
  onSearch: null,
  currentDateInterval: 0,
  language: 'en',
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchFieldContainer);
