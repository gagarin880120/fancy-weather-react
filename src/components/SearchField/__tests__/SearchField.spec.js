import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer, { act } from 'react-test-renderer';
import SearchField from '../SearchField';

describe('SearchField', () => {
  describe('component', () => {
    test('should render without crashing', () => {
      const renderer = new ShallowRenderer();
      renderer.render(<SearchField />);

      const result = renderer.getRenderOutput();

      expect(result).toMatchSnapshot();
    });
  });

  describe('SpeechRecognition', () => {
    const mockSpeechRecognition = jest.fn();
    const start = jest.fn();
    const onresult  = jest.fn();

    mockSpeechRecognition.mockReturnValue({
      start: start,
      onresult : onresult
    });

    window.SpeechRecognition = mockSpeechRecognition;
  });

  describe('Events', () => {
    describe('Input events', () => {
      const onSearch = jest.fn();

      const testRenderer = TestRenderer.create(
        <SearchField
          onSearch={onSearch}
        />
      );

      const input = testRenderer.root.findByProps({ type: 'search' });

      test('onChange should change input\'s value', () => {
        act(() => {
          input.props.onChange({ target: { value: 'minsk' } });
        });

        expect(input.props.value).toEqual('minsk');
      });

      describe('onKeyDown', () => {
        test('should not call onSearch function if keycode !== Enter', () => {
          act(() => {
            input.props.onKeyDown({ key: 's', code: 'KeyS', which: 83 });
          });

          expect(onSearch).not.toHaveBeenCalled();
        });

        test('should not call onSearch function if keycode === Enter, but input don\'t have value', () => {
          act(() => {
            input.props.onChange({ target: { value: '' } });
          });

          act(() => {
            input.props.onKeyDown({ key: 'Enter', code: 'Enter', which: 13 });
          });

          expect(onSearch).not.toHaveBeenCalled();
        });

        test('should call onSearch if keycode === Enter and input have value', () => {
          act(() => {
            input.props.onChange({ target: { value: 'minsk' } });
          });

          act(() => {
            input.props.onKeyDown({ key: 'Enter', code: 'Enter', which: 13 });
          });

          expect(onSearch).toHaveBeenCalled();
        });
      });

    });

    describe('voiceButton events', () => {
      const mockSpeechRecognition = jest.fn();
      const start = jest.fn();
      const onresult  = jest.fn();

      mockSpeechRecognition.mockReturnValue({
        start: start,
        onresult : onresult,
      });

      window.SpeechRecognition = null;
      window.webkitSpeechRecognition = mockSpeechRecognition;

      const recognition = new window.webkitSpeechRecognition;

      const onSearch = jest.fn();

      const testRenderer = TestRenderer.create(
        <SearchField
          onSearch={onSearch}
        />
      );

      const button = testRenderer.root.findByProps({ id: 'voiceButton' });

      describe('onClick', () => {
        test('should call onSearch', () => {
          act(() => {
            button.props.onClick();
          });

          act(() => {
            recognition.onresult({results: ['Minsk']});
          });

          expect(onSearch).toHaveBeenCalled();
        });
      });
    });

    describe('searchButton events', () => {
      const onSearch = jest.fn();

      const testRenderer = TestRenderer.create(
        <SearchField
          onSearch={onSearch}
        />
      );

      const input = testRenderer.root.findByProps({ type: 'search' });
      const button = testRenderer.root.findByProps({ id: 'searchButton' });
      describe('onClick', () => {
        test('if input don\'t have value, should not call onSearch function', () => {
          act(() => {
            button.props.onClick();
          });

          expect(onSearch).not.toHaveBeenCalled();
        });

        test('if input have value, should call onSearch function', () => {
          act(() => {
            input.props.onChange({ target: { value: 'minsk' } });
          });

          act(() => {
            button.props.onClick();
          });

          expect(onSearch).toHaveBeenCalled();
        });
      });
    });
  });
});
