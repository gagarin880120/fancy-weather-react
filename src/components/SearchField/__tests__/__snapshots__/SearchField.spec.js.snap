// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchField component should render without crashing 1`] = `
<div>
  <input
    className="searchInput"
    name="search"
    onChange={[Function]}
    onKeyDown={[Function]}
    type="search"
    value=""
  />
  <button
    className="button"
    id="voiceButton"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fas fa-microphone"
    />
  </button>
  <div
    className="pulsatingCircle"
    style={
      Object {
        "display": "none",
      }
    }
  />
  <button
    className="button"
    id="searchButton"
    onClick={[Function]}
    type="button"
  >
    <i
      className="fas fa-search"
    />
  </button>
</div>
`;
